--- agenda-block.tpl.php	2012-03-21 12:19:46.285911100 +0200
+++ agenda-block3.tpl.php	2012-03-21 12:19:45.341857100 +0200
@@ -19,13 +19,14 @@
   <?php foreach ($events as $day): ?>
   <?php
   $date = $day[0]['start date'];
-
+  $timestamp_date = strtotime($date);
+  $true_date = format_date($timestamp_date, $type = 'medium', $format = $block->dateformat);
   // Substitute today/yesterday/tomorrow
   if (isset($dates[$day[0]['when']])) {
-    $date = $dates[$day[0]['when']];
+    $true_date = $dates[$day[0]['when']];
   }
   ?>
-  <p><?php echo $date; ?></p>
+  <p><?php echo $true_date;  ?></p>
   <ol>
   <?php foreach ($day as $event): ?>
     <li class="cal_<?php echo $event['index']; ?>">
