diff --git a/tvi.module b/tvi.module
index 35cebac..4468792 100755
--- a/tvi.module
+++ b/tvi.module
@@ -198,16 +198,23 @@ function tvi_render_view($tid = '', $depth = 0, $op = 'page') {
 
   list($view, $display, $term, $settings) = tvi_get_view_info($tid);
   if (is_object($view) && $display) {
+    if (isset($view->display_handler->display->display_options["arguments"]["term_node_tid_depth"])) {
+      $args = array($tid, $view->display_handler->display->display_options["arguments"]["term_node_tid_depth"]["depth"]);
+    }
+    else {
+      $args = array($tid);
+    }
+
     if ($view->display[$display]->display_plugin == 'block') {
       // If it's a block display, views returns a block array which won't work
       // as a page callback so we need to explicitly set the page title
       // and just return the $block['content']
-      $block = $view->execute_display($display, array($tid, $view->display_handler->display->display_options["arguments"]["term_node_tid_depth"]["depth"]));
+      $block = $view->execute_display($display, $args);
       drupal_set_title($block['subject']);
       return $block['content'];
     }
     else {
-      return $view->execute_display($display, array($tid, $view->display_handler->display->display_options["arguments"]["term_node_tid_depth"]["depth"]));
+      return $view->execute_display($display, $args);
     }
   }
 
